batch/batch/driver/main.py

```diff
@@ -1322,6 +1330,143 @@ async def monitor_system(app):
     monitor_instances(app)
 
 
+async def compact_agg_billing_project_users_table(db: Database):
+    @transaction(db)
+    async def compact(tx: Transaction, target: dict):
+        original_usage = await tx.execute_and_fetchone(
+            '''
+SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+FOR UPDATE;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        await tx.just_execute(
+            '''
+DELETE FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s AND token != 0;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        await tx.execute_update(
+            '''
+INSERT INTO aggregated_billing_project_user_resources_v3 (billing_project, `user`, resource_id, token, `usage`)
+VALUES (%s, %s, %s, %s, %s)
+ON DUPLICATE KEY UPDATE `usage` = %s
+''',
+            (
+                target['billing_project'],
+                target['user'],
+                target['resource_id'],
+                0,
+                original_usage['usage'],
+                original_usage['usage'],
+            ),
+        )
+
+        new_usage = await tx.execute_and_fetchone(
+            '''
+SELECT CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+FROM aggregated_billing_project_user_resources_v3
+WHERE billing_project = %s AND `user` = %s AND resource_id = %s
+GROUP BY billing_project, `user`, resource_id;
+''',
+            (target['billing_project'], target['user'], target['resource_id']),
+        )
+
+        if new_usage['usage'] != original_usage['usage']:
+            raise ValueError(
+                f'problem in audit for {target}. original usage = {original_usage} but new usage is {new_usage}. aborting'
+            )
+
+    targets = db.execute_and_fetchall(
+        '''
+SELECT billing_project, `user`, resource_id, COUNT(*) AS n_tokens
+FROM aggregated_billing_project_user_resources_v3
+WHERE token != 0
+GROUP BY billing_project, `user`, resource_id
+ORDER BY n_tokens DESC
+LIMIT 10000;
+''',
+        query_name='find_agg_billing_project_user_resource_to_compact',
+    )
+
+    async for target in targets:
+        await compact(target)  # pylint: disable=no-value-for-parameter
```

Comment:
I don't think we should hold the database connection open for the limit-10000 query. That's gonna stream the records 100 at a time from the DB. In the meantime, it keeps the query running and the connection open. I worry we are at risk of intermittent connection issues. I also worry about deadlocking since the DB needs to maintain for the limit-10000 query a consistent view of the aggregation table. That table is usually frequently mutated when batch is working! Finally, I feel a bit weird about mutating the table in one query that you're reading in a another query.

Reply:
I changed it to get all of the targets in a list before iterating. The deadlocking isn't an issue because we're not holding locks to those tables, we don't care if the list of targets is outdated, and if the target doesn't need to be compacted, then it will just be a no-op when the function is called. But I agree we shouldn't hold the connection the entire time the compaction loop runs.